all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+    self.per_worker_allowed.acquire().await?;
+    Ok(())
+  }
+  
+  fn try_reserve_slot(&self) -> Result<bool, anyhow::Error> { /* ... */ }
+  
+  fn mark_slot_used(&self, info: &ActivitySlotInfo) {
+    if info.activity_type == "there_can_only_be_one" {
+      // Probably kicks off something in the background, but immediately blocks slot reservation
+      self.lock_svc_client.eventually_acquire_lock();
+    }
+  }
+  
+  fn release_slot(&self, info: &ActivitySlotInfo) {
+    if info.activity_type == "there_can_only_be_one" {
+      self.lock_svc_client.release_lock();
+    }
+    self.per_worker_allowed.release();
+  }
+  
+  fn available_slots(&self) -> Option<usize> { Some(self.per_worker_allowed.available_permits()) }
+}
+```
+
+## Drawbacks
+
+* Gives users a big footgun. This is mitigated by the fact that most users are probably just going to use one of our
+  default implementations. Inevitably though, someone will implement a custom slot supplier or cache manager with buggy
+  logic and contact support about it.
+* More complexity in the core/lang bridge, nothing too bad though.
+
+## Alternatives
+
+We could not expose this as a user-implementable interface at all and only offer a handful of default implementations.
+The benefits of user customization are substantial though, and seem to outweigh the mild complexity increase. Users will
+always know more about their workload than we can.
+
+Specifically concerning `mark_slot_used`, we could potentially allow an error to be returned here, causing the SDK to
+immediately fail the task instead of working on it. This is maybe a nice way to let users reject tasks they don't want
+this worker to handle... however there are some obvious potential thrashing issues there (ex: only one worker rejecting
+the same task over and over). Allowing this seems like it'd mostly be a way for users to bandaid over something that was
+really a mistake in the way their task queues are architected.
+
+## Adoption & Teaching
+
+We should roll out the feature by introducing the interface & reimplementing the existing behavior behind that
+interface. There will be no visible default behavior changes. Likely we'll include some other OOTB implementations like
+a memory-based one that users can opt-in to.
+
+We might, at some point, decide that the default implementation should change from the fixed-number-of-slots impl to
+a more dynamic one, if we find that it substantially outperforms in most scenarios. If and when we do make that change,
+we'll want to give users some advance warning, since although the change is not technically breaking, it could have 
+knock-on effects in terms of the external load on the systems their workers are interacting with.
+
+Docs & education will both need content about what the interface is for, what default implementations are provided &
+what their limitations are, and how to implement a custom one.
+
+## Open questions
+
+### Async/fallible `mark_slot_used` and `release_slot`
```

Comment:
Is this questions relevant to all SDKs or just a Rust SDK implementation of a `SlotSupplier`?

Reply:
All of them